File: js/src/twitter/createVerifiedTwitterRegistry.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createVerifiedTwitterRegistry = void 0;
const web3_js_1 = require("@solana/web3.js");
const constants_1 = require("../constants");
const createInstruction_1 = require("../instructions/createInstruction");
const state_1 = require("../state");
const getHashedNameSync_1 = require("../utils/getHashedNameSync");
const getNameAccountKeySync_1 = require("../utils/getNameAccountKeySync");
const int_1 = require("../int");
const createReverseTwitterRegistry_1 = require("./createReverseTwitterRegistry");
// Signed by the authority, the payer and the verified pubkey
function createVerifiedTwitterRegistry(connection, twitterHandle, verifiedPubkey, space, // The space that the user will have to write data into the verified registry
payerKey) {
    return __awaiter(this, void 0, void 0, function* () {
        // Create user facing registry
        const hashedTwitterHandle = (0, getHashedNameSync_1.getHashedNameSync)(twitterHandle);
        const twitterHandleRegistryKey = (0, getNameAccountKeySync_1.getNameAccountKeySync)(hashedTwitterHandle, undefined, constants_1.TWITTER_ROOT_PARENT_REGISTRY_KEY);
        const lamports = yield connection.getMinimumBalanceForRentExemption(space + state_1.NameRegistryState.HEADER_LEN);
        let instructions = [
            (0, createInstruction_1.createInstruction)(constants_1.NAME_PROGRAM_ID, web3_js_1.SystemProgram.programId, twitterHandleRegistryKey, verifiedPubkey, payerKey, hashedTwitterHandle, new int_1.Numberu64(lamports), new int_1.Numberu32(space), undefined, constants_1.TWITTER_ROOT_PARENT_REGISTRY_KEY, constants_1.TWITTER_VERIFICATION_AUTHORITY),
        ];
        instructions = instructions.concat(yield (0, createReverseTwitterRegistry_1.createReverseTwitterRegistry)(connection, twitterHandle, twitterHandleRegistryKey, verifiedPubkey, payerKey));
        return instructions;
    });
}
exports.createVerifiedTwitterRegistry = createVerifiedTwitterRegistry;

File: python/src/twitter/create_verified_twitter_registry.py
from dis import Instruction
from typing import List

from solana.constants import SYSTEM_PROGRAM_ID
from solana.rpc.async_api import AsyncClient
from solders.pubkey import Pubkey

from NameRegistryState import NameRegistryState
from constants import NAME_PROGRAM_ID, TWITTER_ROOT_PARENT_REGISTRY_KEY, TWITTER_VERIFICATION_AUTHORITY
from instructions import create_instruction
from int import Numberu64, Numberu32
from twitter.create_reverse_twitter_registry import create_reverse_twitter_registry
from twitter.get_twitter_registry_key import get_twitter_registry_key
from utils import get_hashed_name, get_name_account_key


async def create_verified_twitter_registry(
        connection: AsyncClient,
        twitter_handle: str,
        verified_pubkey: Pubkey,
        space: int,
        payer: Pubkey,
) -> List[Instruction]:
    hashed_twitter_handle = get_hashed_name(twitter_handle)
    twitter_handle_registry = get_name_account_key(
        hashed_twitter_handle,
        None,
        TWITTER_ROOT_PARENT_REGISTRY_KEY
    )

    lamports = await connection.get_minimum_balance_for_rent_exemption(
        space + NameRegistryState.HEADER_LEN
    )
    if lamports is None:
        raise Exception("Failed to get minimum balance for rent exemption")

    ixs = [
        create_instruction(
            NAME_PROGRAM_ID,
            SYSTEM_PROGRAM_ID,
            twitter_handle_registry,
            verified_pubkey,
            payer,
            hashed_twitter_handle,
            Numberu64(lamports.value),
            Numberu32(space),
            None,
            TWITTER_ROOT_PARENT_REGISTRY_KEY,
            TWITTER_VERIFICATION_AUTHORITY
        )
    ]

    ix_create_registry = await create_reverse_twitter_registry(
        connection,
        twitter_handle,
        twitter_handle_registry,
        verified_pubkey,
        payer
    )
    ixs.extend(ix_create_registry)
    return ixs

